wormhole.c

```diff
@@ -94,31 +110,98 @@ static bool wormhole_init(struct tenstorrent_device *tt_dev) {
 	return false;
 }
 
+
+#define ARC_NOC_X 0
+#define ARC_NOC_Y 10
+static int wormhole_hwmon_read(u64 offset, struct tt_hwmon_context *context, int channel, long *val) {
+	struct tenstorrent_device *tt_dev = context->tt_dev;
+	struct wormhole_device *wh_dev = tt_dev_to_wh_dev(tt_dev);
+	struct eth_addr_t *remote_addr;
+	u32 eth_channel;
+	u8 lock_index;
+	bool ok;
+
+	offset += context->telemetry_offset;
+
+	if (channel == 0) {
+		offset += ARC_CSM_START;
+		*val = ioread32(wh_dev->bar4_mapping + offset);
+		return 0;
+	}
+
+	if (wh_dev->num_connected_cores == 0)
+		return -EOPNOTSUPP;
+
+	// Need to do a remote read, so adjust the offset - the ARC will be accessed
+	// via the NOC rather than via PCI BAR.
+	offset += ARC_CSM_NOC;
+
+	// Use the first connected core to determine the remote chip's address.
+	remote_addr = &wh_dev->connected_eth_cores[0].remote;
+
+	// Use eth core 0 to do the actual read.
+	eth_channel = 0;
+	lock_index = TENSTORRENT_LOCK_ETH(eth_channel);
+
+	// FIXME: UMD must use the KMD locking API to avoid conflicts here.
```

Comment:
I need to address this in UMD before this gets merged.

Reply:
There was a simple solution to this (use a core that UMD does not touch) that won't be viable due to an upcoming change for Metal: ETH cores that are not connected to a peer can not be relied on to have traditional firmware.  Although the first 8KB of L1 will be the same so logic to discover the topology will still work, software can't use any arbitrary core for remote transfers.

Reply:
Is there a solution for safe queue submission? Other than this issue, I think this is ready to go.

Reply:
I wonder if we can track whether we've seen that API being used by the processes that have the device open. If they've never used it, maybe we should just stay away and not access the remote. I.e. return -EBUSY where appropriate -- unless that triggers a retry/wait loop up the call chain that wouldn't complete.

Reply:
@alewycky-tenstorrent - the only way is a locking scheme (or don't share the core).
@olofj - interesting idea - I'll try that

Reply:
@olofj, a first pass at implementing your idea is at 9c64bf82be55126aafa61d1bc7706d3f34964f96.

Here is what `sensors` looks like when the driver is returning -EBUSY in this case:
```
wormhole-pci-6100
Adapter: PCI adapter
asic1_vcore:   720.00 mV (max =  +0.95 V)
ERROR: Can't get value of subfeature in1_max: Can't read
asic2_vcore:        N/A  (max =  +0.00 V)
asic1_temp:     +47.2°C  (high = +75.0°C)
ERROR: Can't get value of subfeature temp2_max: Can't read
asic2_temp:         N/A  (high =  +0.0°C)
asic1_power:    12.00 W  (max =  85.00 W)
ERROR: Can't get value of subfeature power2_max: Can't read
asic2_power:        N/A  (max =   0.00 W)
asic1_current:  17.00 A  (max = +160.00 A)
ERROR: Can't get value of subfeature curr2_max: Can't read
asic2_current:      N/A  (max =  +0.00 A)
```

Reply:
Improved implementation at 1e374252f14eaaa4f59e99c452b14078dd75ef28.
As an aside, hwmon indexing starts at 1 for everything but voltage (spelled `in`), which starts at 0 -- hence `in1_max` vs `temp2_max`, `power2_max`, etc. above.